src/System.Memory/src/System/SpanSortHelpers.Keys.Comparison.cs

```diff
@@ -0,0 +1,329 @@
+
+                if (left >= right)
+                    break;
+
+                Swap(ref keys, left, right);
+            }
+            // Put pivot in the right location.
+            right = hi - 1;
+            if (left != right)
+            {
+                Swap(ref keys, left, right);
+            }
+            return left;
+        }
+
+        private static void HeapSort<TKey>(
+            ref TKey keys, int lo, int hi,
+            Comparison<TKey> comparison)
+            
+        {
+            Debug.Assert(comparison != null);
+            Debug.Assert(lo >= 0);
+            Debug.Assert(hi > lo);
+
+            int n = hi - lo + 1;
+            for (int i = n / 2; i >= 1; --i)
+            {
+                DownHeap(ref keys, i, n, lo, comparison);
+            }
+            for (int i = n; i > 1; --i)
+            {
+                Swap(ref keys, lo, lo + i - 1);
+                DownHeap(ref keys, 1, i - 1, lo, comparison);
+            }
+        }
+
+        private static void DownHeap<TKey>(
+            ref TKey keys, int i, int n, int lo,
+            Comparison<TKey> comparison)
+            
+        {
+            Debug.Assert(comparison != null);
+            Debug.Assert(lo >= 0);
+
+            //TKey d = keys[lo + i - 1];
+            ref TKey keysAtLo = ref Unsafe.Add(ref keys, lo);
+            ref TKey keysAtLoMinus1 = ref Unsafe.Add(ref keysAtLo, -1); // No Subtract??
+            TKey d = Unsafe.Add(ref keysAtLoMinus1, i);
+            var nHalf = n / 2;
+            while (i <= nHalf)
+            {
+                int child = i << 1;
+
+                //if (child < n && comparison(keys[lo + child - 1], keys[lo + child]) < 0)
+                if (child < n &&
+                    comparison(Unsafe.Add(ref keysAtLoMinus1, child), Unsafe.Add(ref keysAtLo, child)) < 0)
+                {
+                    ++child;
+                }
+
+                //if (!(comparison(d, keys[lo + child - 1]) < 0))
+                if (!(comparison(d, Unsafe.Add(ref keysAtLoMinus1, child)) < 0))
+                    break;
+
+                // keys[lo + i - 1] = keys[lo + child - 1]
+                Unsafe.Add(ref keysAtLoMinus1, i) = Unsafe.Add(ref keysAtLoMinus1, child);
+
+                i = child;
+            }
+            //keys[lo + i - 1] = d;
+            Unsafe.Add(ref keysAtLoMinus1, i) = d;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void InsertionSort<TKey>(
+            ref TKey keys, int lo, int hi,
+            Comparison<TKey> comparison)
+            
+        {
+            Debug.Assert(lo >= 0);
+            Debug.Assert(hi >= lo);
+
+            for (int i = lo; i < hi; ++i)
+            {
+                int j = i;
+                //t = keys[i + 1];
+                var t = Unsafe.Add(ref keys, j + 1);
+                // TODO: Would be good to be able to update local ref here
```

Comment:
Can you please file an issue instead of adding the inline TODOs. They are harder to track (especially long term).